lib/Runtime/Library/JSON.cpp

```diff
@@ -647,42 +679,68 @@ namespace JSON
             }
             else
             {
-                uint32 precisePropertyCount = 0;
                 Js::JavascriptStaticEnumerator enumerator;
                 if (object->GetEnumerator(&enumerator, EnumeratorFlags::SnapShotSemantics | EnumeratorFlags::EphemeralReference, scriptContext))
                 {
-                    bool isPrecise;
-                    uint32 propertyCount = GetPropertyCount(object, &enumerator, &isPrecise);
-                    if (isPrecise)
-                    {
-                        precisePropertyCount = propertyCount;
-                    }
+                      Js::DynamicObject * dynamicObject = ((Js::DynamicObject*)object);
+                      if (ReplacerFunction != replacerType || !dynamicObject->HasObjectArray())
+                      {
+                        int propertyCount = dynamicObject->GetPropertyCount();
 
-                    if (ReplacerFunction != replacerType)
-                    {
-                        // Reserve initial slots for properties. +1 stands for extra property being pushed during the stringify
-                        result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount + 1);
+                        result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount);
+                        Js::DynamicTypeHandler * typeHandler = dynamicObject->GetTypeHandler();
                         enumerator.Reset();
-                        while ((propertyName = enumerator.MoveAndGetNext(id)) != NULL)
+                        Js::PropertyId previousId;
+                        if (dynamicObject->HasObjectArray())
                         {
-                             if (id == Js::Constants::NoProperty)
+                            int totalNumPropertyCount = this->GetPropertyCount(object, &enumerator) - propertyCount;
+                            Assert(totalNumPropertyCount > 0);
+                            int index = 0;
+                            enumerator.Reset();
+                            while ((propertyName = enumerator.MoveAndGetNext(id)) != NULL && index < totalNumPropertyCount)
                             {
-                                //if unsuccessful get propertyId from the string
+                                // if unsuccessful get propertyId from the string
```

Comment:
Is this correct? Previously this was under a `if (id == Js::Constants::NoProperty)` conditional, and so the comment/behaviour made sense as a fallback, but now it always gets the property record.

Reply:
@MSLaguana there are differences in both cases. In this particular case, I don't know any reason for checking `id == Js::Constants::NoProperty`.